ipa-core/src/protocol/ipa_prf/mod.rs

```diff
@@ -202,6 +205,92 @@ where
     }
 }
 
+impl<BK, TV, TS> OPRFIPAInputRow<BK, TV, TS>
+where
+    BK: BooleanArray,
+    TV: BooleanArray,
+    TS: BooleanArray,
+{
+    fn join_fields(
+        match_key: MatchKey,
+        is_trigger: Boolean,
+        breakdown_key: BK,
+        trigger_value: TV,
+        timestamp: TS,
+    ) -> <Self as shuffle::Shuffleable>::Share {
+        let mut share = <Self as shuffle::Shuffleable>::Share::ZERO;
+
+        BooleanArrayWriter::new(&mut share)
+            .write(&match_key)
+            .write_boolean(is_trigger)
+            .write(&breakdown_key)
+            .write(&trigger_value)
+            .write(&timestamp);
+
+        share
+    }
+
+    fn split_fields(
+        share: &<Self as shuffle::Shuffleable>::Share,
+    ) -> (MatchKey, Boolean, BK, TV, TS) {
+        let bits = BooleanArrayReader::new(share);
+        let (match_key, bits) = bits.read();
+        let (is_trigger, bits) = bits.read_boolean();
+        let (breakdown_key, bits) = bits.read();
+        let (trigger_value, bits) = bits.read();
+        let (timestamp, _) = bits.read();
+        (
+            match_key,
+            is_trigger,
+            breakdown_key,
+            trigger_value,
+            timestamp,
+        )
+    }
+}
+
+impl<BK, TV, TS> shuffle::Shuffleable for OPRFIPAInputRow<BK, TV, TS>
+where
+    BK: BooleanArray,
+    TV: BooleanArray,
+    TS: BooleanArray,
+{
+    type Share = BA112;
```

Comment:
this only works for specific TV and TS, is there a place for at least a runtime check? Or implementing it only for `OPRFIPAInputRow<BA8, BA3, BA20>`?

Reply:
It would have failed at runtime within the packing routines, but since that may not give a clear message, I'll add a debug assertion that the share type is big enough. (I pushed this to #1397 since it goes with those changes.)

Since we support two `BK` types, we'd need `Shuffleable` implementations for two sets of type parameter values for each of the shuffle input types. Which is doable with macros, and I think we'd need to go that route if we wanted a compile-time check.